backend/service/k8s/k8s.go

```diff
@@ -56,6 +58,7 @@ type Service interface {
 	DescribePod(ctx context.Context, clientset, cluster, namespace, name string) (*k8sapiv1.Pod, error)
 	DeletePod(ctx context.Context, clientset, cluster, namespace, name string) error
 	ListPods(ctx context.Context, clientset, cluster, namespace string, listPodsOptions *k8sapiv1.ListPodsOptions) ([]*k8sapiv1.Pod, error)
+	UpdatePodAnnotations(ctx context.Context, clientset, cluster, namespace, name string, expectedAnnotations, newAnnotations Annotations) error
```

Comment:
it would be good if we could come up with a common and extensible way of handling metadata updates. this API works much different than UpdateDeployment for example.

i think the precondition functionality is great so wondering if we can somehow unify the way all of these work.

cc @mcutalo88 for thoughts

Reply:
I was just going to comment on the precondition and the label / annotation update process in general. I dont really see the value in doing a precondition check given that updating a map is easily unit testable at least in this case Im not seeing the value and it looks too complex for the given task.

Ideally I would prefer us to use a `Patch()` here since we should be merging the new and old objects together, I realized that the original intention was to delete all annotations that were not supplied by the caller I left a comment below explaining why we should not do that.

Let me digest this PR a bit more.

Reply:
The goal of the preconditions is to prevent some race conditions or updates using stale views.

The updates will always be applied if no preconditions are specified, but if preconditions are specified the client will first get the pod and check whether the preconditions match before applying the updates.

In the case of the Squeeze workflow it prevents clutch from overwriting load balancing weight updates in scenarios like the following:

a. User opens multiple edit tabs, which show the current load balancing weight and then tries to change the weight with both of the tabs.
b. User A opens the edit tab and sees weight X. User B opens the edit tab, sees weight X and sets weight to Y. User A still seeing weight X tries to update weight to Z, which would overwrite the change that user B did.

Reply:
Thanks for the clarification @gkleiman I see what you'r trying to accomplish now.